client/src/main/java/io/pravega/client/control/impl/ControllerImpl.java

```diff
@@ -1282,30 +1307,78 @@ private StreamSegments getStreamSegments(final SegmentRanges ranges, long reques
+        Segment segment = Segment.fromScopedName(qualifiedSegmentName);
+        CachedPravegaNodeUri nodeUri = getSegmentEndpointFromCache(segment);
+        //Read from cache if the segment endpoint already exists and the refresh interval is not expired.
+        if (nodeUri != null && nodeUri.getTimer().getElapsedMillis() <= CachedPravegaNodeUri.MAX_BACKOFF_MILLIS) {
+            log.info("Fetching the endpoint details of segment {} from cache", segment.getSegmentId());
+            return nodeUri.getPravegaNodeUri();
+        } else if (nodeUri != null && nodeUri.getTimer().getElapsedMillis() > CachedPravegaNodeUri.MAX_BACKOFF_MILLIS) {
+            // Trigger a background call and refresh the cache.
+            endPointCacheMap.put(segment, new CachedPravegaNodeUri(new Timer(), getPravegaNodeUri(segment)));
+            return nodeUri.getPravegaNodeUri();
+        } else {
+            CompletableFuture<PravegaNodeUri> nodeInfo = getPravegaNodeUri(segment);
+            endPointCacheMap.put(segment, new CachedPravegaNodeUri(new Timer(), nodeInfo));
+            return nodeInfo;
+        }
+    }
+
+    @Override
+    public void updateStaleValueInCache(String segmentName, PravegaNodeUri errNodeUri) {
+        Exceptions.checkNotNullOrEmpty(segmentName, "segmentName");
+        Segment segment = Segment.fromScopedName(segmentName);
         final long requestId = requestIdGenerator.get();
-        long traceId = LoggerHelpers.traceEnter(log, "getEndpointForSegment", qualifiedSegmentName, requestId);
+        long traceId = LoggerHelpers.traceEnter(log, "updateStaleValueInCache", segmentName, errNodeUri, requestId);
+        CachedPravegaNodeUri cachedNode = getSegmentEndpointFromCache(segment);
+        if (cachedNode != null) {
+            cachedNode.getPravegaNodeUri().thenAccept(cachedNodeUri -> {
+                if (cachedNodeUri.getEndpoint().equals(errNodeUri.getEndpoint()) && cachedNodeUri.getPort() == errNodeUri.getPort()) {
+                    // enforce cache refresh in case of stale value
+                    log.debug(requestId, "Refreshing stale value in cache for segment {}", segment.getSegmentId());
+                    endPointCacheMap.put(segment, new CachedPravegaNodeUri(new Timer(), getPravegaNodeUri(segment)));
+                }
+            }).whenComplete((x, e) -> {
+                if (e != null) {
+                    log.warn("updateStaleValueInCache failed for segment {}: ", segmentName, e);
+                    endPointCacheMap.remove(segment);
+                }
+                LoggerHelpers.traceLeave(log, "updateStaleValueInCache", traceId, requestId);
+            });
+        }
+    }
 
+    //Avoid invoking getPravegaNodeUri directly.Invoke getEndpointForSegment API instead, to get the endpoint details of a segment.
+    // It is fast comparatively as it cache the endpoint details.
+    @VisibleForTesting
+    public CompletableFuture<PravegaNodeUri> getPravegaNodeUri(Segment segment) {
```

Comment:
Can this be made default visibility? ControllerImplTest and ConnectionPoolingTest are in the same package.

Reply:
We are invoking this in a system test (CacheSegmentToHostMappingPerformanceTest) , hence made public.